### PR TITLE
Ean/application test zip issues

### PR DESCRIPTION
**Pull request by ean**
@mpolden

### CODE REVIEW

File: client/go/vespa/application.go [hunk 1]
```diff
@@ -153,25 +154,23 @@ func FindApplicationPackage(zipOrDir string, requirePackaging bool) (Application
...
+			return ApplicationPackage{Path: path, TestPath: testPath}, nil
 		}
-		return ApplicationPackage{Path: filepath.Join(zipOrDir, "src", "main", "application")}, nil
+		return ApplicationPackage{Path: path}, nil
 	}
 	if util.PathExists(filepath.Join(zipOrDir, "services.xml")) {
 		return ApplicationPackage{Path: zipOrDir}, nil
 	}
-	return ApplicationPackage{}, errors.New("Could not find an application package source in '" + zipOrDir + "'")
+	return ApplicationPackage{}, fmt.Errorf("Could not find an application package source in '%s'", zipOrDir)
```
Comment (mpolden) [inline client/go/vespa/application.go:224]: Errors should not be capitalized: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: client/go/cmd/prod.go [hunk 2]
```diff
@@ -377,18 +374,29 @@ func prompt(r *bufio.Reader, question, defaultAnswer string, validator func(inpu
...
+	testPath := app.TestPath
+	if app.IsZip() {
+		path, err := app.Unzip(true)
+		if err != nil {
+			return err
+		}
+		defer os.RemoveAll(path)
+		testPath = path
 	}
-	for suite, required := range suites {
```
Comment (mpolden) [inline client/go/cmd/prod.go]: I liked the map approach better as the destructuring here was more clear.

File: client/go/vespa/application.go [hunk 3]
```diff
@@ -146,6 +146,55 @@ func (ap *ApplicationPackage) zipReader(test bool) (io.ReadCloser, error) {
...
+	}
+	tmp, err := os.MkdirTemp(os.TempDir(), "vespa-test-pkg")
+	if err != nil {
+		return "", err
+	}
+	path := ap.Path
+	if test {
+		path = ap.TestPath
+	}
+	f, err := zip.OpenReader(path)
```
Comment (mpolden) [inline client/go/vespa/application.go:161]: `f` should be closed in a `defer`.

Comment (mpolden) [inline client/go/vespa/application.go:161]: If this fails, `tmp` will still exist. How about this:
```
cleanTemp := true
tmp, err := os.MkdirTemp(os.TempDir(), "vespa-test-pkg")
if err != nil {
    return "", err
}
defer func() {
    if cleanTemp {
        os.RemoveAll(tmp)
    }
}()
```

then remove the explicit `os.RemoveAll`s below and simply do `cleanTemp = false` before returning.

File: client/go/cmd/prod.go (hunk 2 above)
Comment (ean) [reply client/go/cmd/prod.go]: agreed, and fixed.

File: client/go/vespa/application.go (hunk 3 above)
Comment (ean) [reply client/go/vespa/application.go:167]: I like it.

**Review by mpolden — APPROVED** [review]
💯

**Pull request merged**